src/shell/shell.cpp

```diff
@@ -70,67 +70,129 @@ DOS_Shell::DOS_Shell()
 	}
 }
 
-void DOS_Shell::GetRedirection(char *line,
-                               std::string &in_file,
-                               std::string &out_file,
-                               std::string &pipe_file,
-                               bool *append)
+// This function gets redirection targets from the given shell command line and
+// returns the results as a struct. The results include the redirection targets
+// as well as the processed command line with the targets stripped off.
+//
+// Note that real MS-DOS is quite nuanced in its whitespace handling:
+// - 'echo 1>out.txt' produces a 3-byte file: '1CRLF'
+// - 'echo 1 > out.txt' produces a 4-byte file: '1 CRLF'
+// - 'echo 1 >out.txt ' produces a 5-byte file: '1  CRLF'
+// This behavior is replicated here; see the shell_redirection_tests for more
+// examples.
+//
+std::optional<DOS_Shell::RedirectionResults> DOS_Shell::GetRedirection(
+        const std::string_view line)
```

Comment:
Nice use of `optional`, I approve 😄

Reply:
It was really useful to support the syntax error scenario after seeing real MS-DOS doesn't process any of the content and just bails out on the command entirely.